File: secret_santa.rb
# Murtaza Jafferji
# JBS Prework
# 5/28/10
#
# Secret Santa
# This program is designed to take the names and email addresses of participants from a user 
#   and then choose a Secret Santa for every person in the list. Santas are selected with the
#   condition that someone's Santa is not part of his/her family. Each Santa is emailed
#   with the name of their future gift recipient. 
# Preconditions: Install tlsmail gem

# a person class to manage participant data
class Person
  attr_accessor :fname, :lname, :email
 
  def initialize(fname, lname, email)
    @fname = fname
    @lname = lname
    @email = email
  end
end
  
puts 'Enter the data (type "done" when finished)'
puts 'FIRST_NAME space FAMILY_NAME space <EMAIL_ADDRESS> newline' 
$people_array = Array.new
flag = true
  while flag
    line = gets.chomp
    if line.downcase == 'done'
      flag = false
    elsif line =~ /\w* \w* <\w*@\w*.\w*>/ #checks if data entered in correct format
      split_line = line.split #splits the line into an array of its components
      email_address = split_line[2].reverse.chop #deletes the "<" and ">" from the email address
      email_address = email_address.reverse.chop
      $people_array.push Person.new(split_line[0], split_line[1], email_address)
    else 
      puts "Invalid entry!"
    end
end

$pairs = $people_array.sort_by { rand } #creates a new array of participants in a random ordering

# returns true if a randomly sorted second array is ordered in such a way that corresponding 
# indices with the first array match two people without the same last name 
def match
  $people_array.each_with_index do |person, i|
    if person.lname.downcase == $pairs[i].lname.downcase
      return false
    end
  end
  return true
end

# randomly sort until an acceptable ordering is found
$pairs = $people_array.sort_by { rand } until match 

# prints each Santa's name with their future gift recipient
$people_array.each_with_index do |person, i|
  puts person.fname + " " + person.lname + " is " + $pairs[i].fname + " " + $pairs[i].lname + "'s Secret Santa."
end
  
# asks user for Gmail information for sending out emails
puts "\nEnter your Gmail username:"
@username = gets.chomp
puts "Enter your Gmail password:"
@password = gets.chomp

# sends an email to each Santa with the name of their future gift recipient
def send_email(santa, recipient)
  require 'tlsmail'  
  require 'time'  
  
content = <<EOF  
From: #{@username}
To: #{santa.email}
Subject: Hey #{santa.fname}! Secret Santa info inside.   
Date: #{Time.now.rfc2822}  
  
You are #{recipient.fname} #{recipient.lname}'s Secret Santa.
EOF
  
  Net::SMTP.enable_tls(OpenSSL::SSL::VERIFY_NONE)  
  Net::SMTP.start('smtp.gmail.com', 587, 'gmail.com', @username, @password, :login) do |smtp|  
    smtp.send_message(content, @username, santa.email)  
  end
end
 
# sends an email to each Santa
$people_array.each_with_index do |person, i|
  send_email(person, $pairs[i])
  
end

File: hw16/SecretSanta.java
import java.util.*;

/* Name: Secret Santa
 * Assignment: hw16
 * Description: This program is designed to take the names and email addresses of participants from a user 
 *   and then choose a Secret Santa for every person in the list. Santas are selected with the
 *   condition that someone's Santa is not part of his/her family. Each Santa is emailed
 *   with the name of their future gift recipient. 
 * Author: Murtaza Jafferj
 * 
 */

// a person class to manage participant data
public class SecretSanta{
	private static List<Person> people_array = new ArrayList<Person>();
	
	public static void ask(){ 
		System.out.println("Enter the data (type \"done\" when finished)");
		System.out.println ("FIRST_NAME space FAMILY_NAME space <EMAIL_ADDRESS> newline"); 
	}

	public static void parse(){
		boolean flag = true;
		while (flag){
			Scanner input = new Scanner(System.in);
			String line = input.nextLine();
			if (line.toLowerCase().equals("done")){
				flag = false;
			}else{
				String[] split_line;
				split_line = line.trim().split("\\s+"); //splits the line into an array of its components
				Person person = new Person(split_line[0], split_line[1], split_line[2]);
				people_array.add(person);
			}
		}
	}
	
	// returns true if a randomly sorted second array is ordered in such a way that corresponding 
	// indices with the first array match two people without the same last name 
	public static boolean match(List<Person> pairs){
		for (int i = 0; i < people_array.size(); i++){
			if (people_array.get(i).lname.toLowerCase() == pairs.get(i).lname.toLowerCase()){
				return false;
			}
		}
		return true;
	}
	
	// randomly sort until an acceptable ordering is found
	public static List<Person> get_result(){
		List<Person> pairs = new ArrayList<Person>();
		for (int i = 0; i < people_array.size(); i++){
			pairs.add(people_array.get(i));
		}
		Collections.shuffle(pairs); //creates a new array of participants in a random ordering
		while (match(pairs) == false){ 
			Collections.shuffle(pairs);
		}
		return pairs;
	}

	// prints each Santa's name with their future gift recipient
	public static void print_pairs(List<Person> pairs){
		for (int i = 0; i < people_array.size(); i++){
			System.out.println(people_array.get(i).fname + " " + people_array.get(i).lname + " is " + pairs.get(i).fname + " " + pairs.get(i).lname + "'s Secret Santa.");
		}
	}

	public static void main(String[] args){
		ask();
		parse();
		print_pairs(get_result());
	}
}